### PR TITLE
OSDOCS-9437: adds audit log config policies MicroShift

### PR DESCRIPTION
**Pull request by ShaunaDiaz**
Version(s):
4.16+

Issue:
[OSDOCS-9437](https://issues.redhat.com/browse/OSDOCS-9437)

Link to docs preview:
[Customizing audit logs](https://75233--ocpdocs-pr.netlify.app/microshift/latest/microshift_configuring/microshift-audit-logs-config.html)

QE review:
- [x] QE has approved this change.

SME review:
- [x] SME has approved this change.

Additional information:
[Release note](https://github.com/openshift/openshift-docs/pull/75867)

### CODE REVIEW

**Comment by ShaunaDiaz** [comment]
@kasturinarra Could you also PTAL at this one? TY!

**Comment by ShaunaDiaz** [comment]
@gangwgr If you want to take a look also?

**Comment by gangwgr** [comment]
@copejon have one query? for default size is 200Mb
```
why it is 100Mb shown here
sudo ls -ltrh /var/log/kube-apiserver/
total 358M
-rw-------. 1 root root 100M May 13 11:51 audit-2024-05-13T15-51-42.132.log
-rw-------. 1 root root 100M May 13 20:41 audit-2024-05-14T00-41-40.725.log
-rw-------. 1 root root 100M May 14 04:30 audit-2024-05-14T08-30-29.480.log
-rw-------. 1 root root 56M May 14 09:24 audit.log
[redhat@dhcp-1-235-245 ~]$ sudo rm audit.log





[7:00](https://redhat-internal.slack.com/archives/D03SQFGHYGK/p1715693429188309)
https://github.com/openshift/openshift-docs/pull/75233/files#diff-5c2cca3b817bccbe26d6a02cd9774a86d959085eaa03f5757817d76fa5573743R21
[7:04](https://redhat-internal.slack.com/archives/D03SQFGHYGK/p1715693645817859)
Added fake logs, it accepted 257Mb
ls -ltrh /var/log/kube-apiserver/
total 559M
-rw-------. 1 root root 100M May 13 11:51 audit-2024-05-13T15-51-42.132.log
-rw-------. 1 root root 100M May 13 20:41 audit-2024-05-14T00-41-40.725.log
-rw-------. 1 root root 100M May 14 04:30 audit-2024-05-14T08-30-29.480.log
-rw-------. 1 root root 257M May 14 09:31 audit-2024-05-14T13-32-11.567.log
-rw-------. 1 root root 1.1M May 14 09:33 audit.log
```

**Comment by gangwgr** [comment]
@copejon I see in doc for writeRequestBodies, we only log for` (create, update, patch, delete, deletecollection) `not for `\"verb\":\"get|list|watch\"`
`In addition to logging metadata for all requests, logs request bodies for every write request to the API servers (create, update, patch, delete, deletecollection). This profile has more resource overhead than the Default profile. [1]`

I see some logs on 4.16.0~rc.1 while doing regression testing

**Comment by copejon** [comment]
Thanks @gangwgr I'll check it out

File: modules/microshift-audit-logs-config-proc.adoc [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+<2> The maximum number of rotated audit log files to retain. After the limit is reached, the log files in order from oldest to newest are deleted until the specified limits are reached. When the value is 0, the limit is disabled. In this example, the value `1` results in only 1 file of size `maxFileSize` being retained in addition to the current active log.
+<3> Specifies the maximum time in days that log files are kept. Files older than this limit will be deleted. When the value is 0, the limit is disabled. In this example, after a log file is more than 7 days old, it is deleted. The deletion happens regardless of whether the live log has reached the maximum file size as give in the `maxFileSize` field value.
+<4> Logs only metadata for read and write requests; does not log request bodies except for OAuth access token requests. If you do not specify this field, the Default profile is used.
+
+. Optional: To specify a specific directory for logs, move the `/var/log/kube-apiserver` directory to your desired location by running the following command:
```
Comment (copejon) [inline modules/microshift-audit-logs-config-proc.adoc:35]:
```suggestion
. Optional: To specify a specific directory for logs, first stop MicroShift, then move the `/var/log/kube-apiserver` directory to your desired location by running the following command:
```

**Review by copejon — COMMENTED** [review]
Just a few notes, otherwise looks good!

**Comment by copejon** [comment]
> @copejon I see in doc for writeRequestBodies, we only log for`(create, update, patch, delete, deletecollection)`not for `\"verb\":\"get|list|watch\"` `In addition to logging metadata for all requests, logs request bodies for every write request to the API servers (create, update, patch, delete, deletecollection). This profile has more resource overhead than the Default profile. [1]`
> 
> I see some logs on 4.16.0~rc.1 while doing regression testing

I'm not able to reproduce this. Let's not block this work on it and I'll follow up with you directly.

**Comment by gangwgr** [comment]
> > @copejon I see in doc for writeRequestBodies, we only log for`(create, update, patch, delete, deletecollection)`not for `\"verb\":\"get|list|watch\"` `In addition to logging metadata for all requests, logs request bodies for every write request to the API servers (create, update, patch, delete, deletecollection). This profile has more resource overhead than the Default profile. [1]`
> > I see some logs on 4.16.0~rc.1 while doing regression testing
> 
> I'm not able to reproduce this. Let's not block this work on it and I'll follow up with you directly.

ok, shared steps https://redhat-internal.slack.com/archives/C03CJTNLKAT/p1715963768827189

**Comment by gangwgr** [comment]
From doc side we are ok, only bug left as on that @copejon working on it. 
/lgtm

File: modules/microshift-audit-logs-config-intro.adoc [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+{microshift-short} includes the following default audit log rotation values:
+
+The `maxFileSize` default is 200Mb.
+The `maxFiles` default is 10 files.
+The `maxFileAge` default is 0, meaning there is no default age limit.
+
+Therefore, the default maximum storage consumption of audit logs is 2000Mb, provided that there are 10 or fewer files.
+
+If you do not specify a value for a field, the default value is used. If you remove a field previously set, the default value is restored after the next {microshift-short} service restart.
```
Comment (agantony) [inline modules/microshift-audit-logs-config-intro.adoc:27]:
```suggestion
{microshift-short} supports the following default values for the audit log rotation:

* `maxFileSize`: The default value is `200`.
* `maxFiles`: The default value is `10`.
* `maxFileAge`: The default value is `0`, which means that there is no default age limit.

Therefore, the maximum storage usage of audit logs is 2000 MB, assuming there are 10 or less files.

If you do not specify a value for a field, the default value is used. If you remove a previously specified field, the default value is restored after the next {microshift-short} service restart.
```

File: modules/microshift-audit-logs-config-proc.adoc [hunk 3]
```diff
@@ -0,0 +1,66 @@
+// Text snippet included in the following assemblies:
+//
+// * microshift_configuring/microshift-audit-logs-config.adoc
+
+:_mod-docs-content-type: PROCEDURE
+[id="microshift-configuring-audit-log-values_{context}"]
+= Configuring audit log values
+
+Use the {microshift-short} service configuration file to implement custom audit log settings.
```
Comment (agantony) [inline modules/microshift-audit-logs-config-proc.adoc:9]:
```suggestion
By using the {microshift-short} service configuration file, you can implement custom settings for the audit log.
```

File: modules/microshift-audit-logs-config-proc.adoc [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+.Procedure
+
+. Add the following `apiServer` stanza to your {microshift-short} `config.yaml` file. If you need to make the file, run the following command:
```
Comment (agantony) [inline modules/microshift-audit-logs-config-proc.adoc:13]:
```suggestion
. To make the file, run the following command:
```

File: modules/microshift-audit-logs-config-proc.adoc [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
++
+[source,terminal]
+----
+$ sudo cat /etc/microshift/config.yaml
+----
++
```
Comment (agantony) [inline modules/microshift-audit-logs-config-proc.adoc:19]:
```suggestion
. Add the following content to your {microshift-short} `config.yaml` file:
+
```

File: modules/microshift-audit-logs-config-proc.adoc [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+<2> The maximum number of rotated audit log files to retain. After the limit is reached, the log files in order from oldest to newest are deleted until the specified limits are reached. When the value is 0, the limit is disabled. In this example, the value `1` results in only 1 file of size `maxFileSize` being retained in addition to the current active log.
+<3> Specifies the maximum time in days that log files are kept. Files older than this limit are deleted. When the value is 0, the limit is disabled. In this example, after a log file is more than 7 days old, it is deleted. The deletion happens regardless of whether the live log has reached the maximum file size as give in the `maxFileSize` field value. File age is determined by the timestamp written in the name of the rotated log file, for example, `audit-2024-05-16T17-03-59.994.log`.
+<4> Logs only metadata for read and write requests; does not log request bodies except for OAuth access token requests. If you do not specify this field, the Default profile is used.
```
Comment (agantony) [inline modules/microshift-audit-logs-config-proc.adoc:35]:
```suggestion
<1> The maximum size in megabytes for the audit log file. If the value is `0`, the limit is disabled. In this example, the file is rotated as soon as the live log reaches the 200 MB limit.
<2> The maximum number of rotated audit log files that you want to retain. After the limit is reached, the log files are deleted in order from oldest to newest until the specified limit is reached. If the value is `0`, the limit is disabled. In this example, the value `1` means that only 1 file of the `maxFileSize" is retained in addition to the current active log.
<3> Specifies the maximum time in days that log files are kept. Files older than this limit are deleted. If the value is `0`, the limit is disabled. In this example, a log file that is older than 7 days is deleted. The deletion takes place regardless of whether the live log has reached the maximum file size specified in the `maxFileSize` field. The file age is determined by the timestamp in the name of the rotated log file, for example, `audit-2024-05-16T17-03-59.994.log`.
<4> Logs only metadata for read and write requests; does not log request bodies, except for OAuth access token requests. If you do not specify this field, the default profile is used.
```

File: modules/microshift-audit-logs-config-proc.adoc [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+. Optional: To specify a specific directory for logs, you can stop {microshift-short}, then move the `/var/log/kube-apiserver` directory to your desired location:
```
Comment (agantony) [inline modules/microshift-audit-logs-config-proc.adoc:37]:
```suggestion
. Optional: To specify a specific directory for logs, you can stop {microshift-short}, and then move the `/var/log/kube-apiserver` directory to your desired location:
```

File: modules/microshift-audit-logs-config-proc.adoc [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+----
+$ sudo systemctl stop microshift
+----
+.. Move the `/var/log/kube-apiserver` directory to your desired location by running the following command:
++
+[source,terminal]
+----
+$ sudo mv /var/log/kube-apiserver <~/kube-apiserver> <1>
+----
+<1> Replace _<~/kube-apiserver>_ with the path to the directory that you want to use.
```
Comment (agantony) [inline modules/microshift-audit-logs-config-proc.adoc:51]:
```suggestion
<1> Specify the path to the directory that you want to use.
```

File: modules/microshift-audit-logs-config-proc.adoc [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+
+. If you specified a specific directory for logs, create a symlink to your custom directory at `/var/log/kube-apiserver` by running the following command:
```
Comment (agantony) [inline modules/microshift-audit-logs-config-proc.adoc:53]: Optional ??

File: modules/microshift-audit-logs-config-proc.adoc [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
++
+[source,terminal]
+----
+$ sudo ln -s <~/kube-apiserver> /var/log/kube-apiserver <1>
+----
+<1> Replace _<~/kube-apiserver>_ with the path to the directory that you want to use. This enables the logs to be collected in sos reports.
+
+. If you are configuring audit log policies on a running instance, restart {microshift-short} by entering the following command:
```
Comment (agantony) [inline modules/microshift-audit-logs-config-proc.adoc:61]: Optional ??

**Review by agantony — COMMENTED** [review]
Added a few comments for your consideration; overall lgtm!

File: microshift_configuring/microshift-audit-logs-config.adoc [hunk 11]
```diff
@@ -0,0 +1,18 @@
+:_mod-docs-content-type: ASSEMBLY
+[id="microshift-audit-logs-config"]
+= Customizing audit logging policies
+include::_attributes/attributes-microshift.adoc[]
+:context: microshift-audit-logs-config
+
+toc::[]
+
+You can use configuration values to control audit log file rotation and retention.
```
[the remark this answers is not in the document]
Comment (ShaunaDiaz) [reply microshift_configuring/microshift-audit-logs-config.adoc:9]: Even better: You can control audit log file rotation and retention by using configuration values.

File: modules/microshift-audit-logs-config-proc.adoc (hunk 3 above)
Comment (ShaunaDiaz) [reply modules/microshift-audit-logs-config-proc.adoc:9]: Better: You can configure audit log settings by using the {microshift-short} service configuration file.

File: modules/microshift-audit-logs-config-proc.adoc (hunk 9 above)
Comment (ShaunaDiaz) [reply modules/microshift-audit-logs-config-proc.adoc:53]: Not _really_. We want logs to be accessible to sos reports and this is how we achieve that.

File: modules/nodes-nodes-audit-config-about.adoc [hunk 12]
```diff
@@ -24,18 +37,33 @@ Audit log profiles define how to log requests that come to the OpenShift API ser
...
 [.small]
 --
-1. Sensitive resources, such as `Secret`, `Route`, and `OAuthClient` objects, are only ever logged at the metadata level. OpenShift OAuth server events are only ever logged at the metadata level.
+1. Sensitive resources, such as `Secret`, `Route`, and `OAuthClient` objects, are only logged at the metadata level.
+ifndef::microshift[]
+OpenShift OAuth server events are only ever logged at the metadata level.
+endif::microshift[]
 --
+ifndef::microshift[]
+By default, {product-title} uses the `Default` audit log profile. You can use another audit policy profile that also logs request bodies, but be aware of the increased resource usage such CPU, memory, and I/O.
```
[the remark this answers is not in the document]
Comment (ShaunaDiaz) [reply modules/nodes-nodes-audit-config-about.adoc:64]: ooo good catch!

**Comment by ShaunaDiaz** [comment]
@agantony Great review, thanks!

File: modules/microshift-audit-logs-config-proc.adoc [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+
+:_mod-docs-content-type: PROCEDURE
+[id="microshift-configuring-audit-log-values_{context}"]
+= Configuring audit log values
+
+You can configure audit log settings by using the {microshift-short} service configuration file.
+
+.Procedure
+
+. Make a copy of the provided `config.yaml.default` file in the `/etc/microshift/` directory, renaming it `config.yaml`. Keep the new {microshift-short} `config.yaml` you create in the `/etc/microshift/` directory. The new `config.yaml` is read at system start and takes precedence over built-in settings.
```
[the remark this answers is not in the document]
Comment (ShaunaDiaz) [reply modules/microshift-audit-logs-config-proc.adoc:13]: Yes, it means whenever the MicroShift starts, either for the first time or as a restart.

**Pull request merged**